server/interceptor/interceptor_test.go

```diff
@@ -0,0 +1,425 @@
+	adminTX.EXPECT().GetTree(gomock.Any(), unknownTreeID).AnyTimes().Return(nil, errors.New("not found"))
+	adminTX.EXPECT().Close().AnyTimes().Return(nil)
+	adminTX.EXPECT().Commit().AnyTimes().Return(nil)
+
+	tests := []struct {
+		desc       string
+		req        interface{}
+		fullMethod string
+		handlerErr error
+		wantErr    bool
+		wantTree   *trillian.Tree
+	}{
+		{
+			desc:       "rpcWithoutTree",
+			req:        &trillian.CreateTreeRequest{},
+			fullMethod: "/trillian.TrillianAdmin/CreateTree",
+		},
+		{
+			desc:       "adminRPC",
+			req:        &trillian.GetTreeRequest{TreeId: logTree.TreeId},
+			fullMethod: "/trillian.TrillianAdmin/GetTree",
+			wantTree:   &logTree,
+		},
+		{
+			desc:       "logRPC",
+			req:        &trillian.GetLatestSignedLogRootRequest{LogId: logTree.TreeId},
+			fullMethod: "/trillian.TrillianLog/GetLatestSignedLogRoot",
+			wantTree:   &logTree,
+		},
+		{
+			desc:       "mapRPC",
+			req:        &trillian.GetSignedMapRootRequest{MapId: mapTree.TreeId},
+			fullMethod: "/trillian.TrillianMap/GetSignedMapRoot",
+			wantTree:   &mapTree,
+		},
+		{
+			desc:       "unknownRequest",
+			req:        "not-a-request",
+			fullMethod: "/trillian.TrillianLog/UnmappedRequest",
+			wantErr:    true,
+		},
+		{
+			desc:       "unknownTree",
+			req:        &trillian.GetTreeRequest{TreeId: unknownTreeID},
+			fullMethod: "/trillian.TrillianAdmin/GetTree",
+			wantErr:    true,
+		},
+	}
+
+	ctx := context.Background()
+	intercept := TreeInterceptor{Admin: admin}
+	for _, test := range tests {
+		handler := &fakeHandler{resp: "handler response", err: test.handlerErr}
+
+		resp, err := intercept.UnaryInterceptor(ctx, test.req, &grpc.UnaryServerInfo{FullMethod: test.fullMethod}, handler.run)
+		if hasErr := err != nil && err != test.handlerErr; hasErr != test.wantErr {
+			t.Errorf("%v: UnaryInterceptor() returned err = %q, wantErr = %v", test.desc, err, test.wantErr)
+			continue
+		} else if hasErr {
+			continue
+		}
+
+		if want := 1; handler.calls != want {
+			t.Errorf("%v: handler called %v times, want = %v", test.desc, handler.calls, want)
+			continue
+		}
+		if handler.resp != resp {
+			t.Errorf("%v: resp = %v, want = %v", test.desc, resp, handler.resp)
+		}
+		if handler.err != err {
+			t.Errorf("%v: err = %v, want = %v", test.desc, err, handler.err)
+		}
+
+		if test.wantTree != nil {
+			switch tree, ok := trees.FromContext(handler.ctx); {
+			case !ok:
+				t.Errorf("%v: tree not in handler ctx", test.desc)
+			case !proto.Equal(tree, test.wantTree):
+				diff := pretty.Compare(tree, test.wantTree)
+				t.Errorf("%v: post-FromContext diff:\n%v", test.desc, diff)
+			}
+		}
+	}
+}
+
+func TestGetRPCInfo(t *testing.T) {
+	tests := []struct {
+		req                      interface{}
+		fullMethod               string
+		wantID                   int64
+		wantType                 trillian.TreeType
+		wantNoTree, wantReadonly bool
+	}{
+		// TrillianAdmin
+		{
```

Comment:
I add all of them because it has a sequence of ack + multiple cursors + ? + profit. We don't have to map all methods, I believe all the logic is already covered. Unless if breaks, ofc, but then adding a new test is a given.

I can remove some methods to make this clearer, if you like.